python/ray/rllib/ppo/ppo_tf_policy.py

```diff
@@ -0,0 +1,175 @@
+        self._inputs["advantages"] = tf.placeholder(
+            tf.float32, name="advantages", shape=(None,))
+        self._inputs["actions"] = ModelCatalog.get_action_placeholder(
+            action_space)
+        # Log probabilities from the policy before the policy update.
+        self._inputs["logprobs"] = tf.placeholder(
+            tf.float32, name="logprobs", shape=(None, logit_dim))
+        # Value function predictions before the policy update.
+        self._inputs["vf_preds"] = tf.placeholder(
+            tf.float32, name="vf_preds", shape=(None,))
+        # KL Coefficient
+        self.loss_in = list(self._inputs.items())
+        self.kl_coeff = tf.placeholder(
+            name="newkl", shape=(), dtype=tf.float32)
+
+    def _setup_graph(self, action_space):
+        self.dist_cls, self.logit_dim = ModelCatalog.get_action_dist(
+            action_space)
+        self.logits = ModelCatalog.get_model(
+            self._inputs["obs"], self.logit_dim, self.config["model"]).outputs
+        self.curr_dist = self.dist_cls(self.logits)
+        self.sampler = self.curr_dist.sample()
+        if self.config["use_gae"]:
+            vf_config = self.config["model"].copy()
+            # Do not split the last layer of the value function into
+            # mean parameters and standard deviation parameters and
+            # do not make the standard deviations free variables.
+            vf_config["free_log_std"] = False
+            with tf.variable_scope("value_function"):
+                self.value_function = ModelCatalog.get_model(
+                    self._inputs["obs"], 1, vf_config).outputs
+            self.value_function = tf.reshape(self.value_function, [-1])
+        else:
+            self.value_function = tf.constant("NA")
+
+    def extra_compute_action_fetches(self):
+        return {"vf_preds": self.value_function, "logprobs": self.logits}
+
+    def extra_apply_grad_fetches(self):
+        return {"kl": self.loss_obj.mean_kl}
+
+    def extra_apply_grad_feed_dict(self):
+        return {self.kl_coeff: self.kl_coeff_val}
+
+    def update_kl(self, sampled_kl):
+        if sampled_kl > 2.0 * self.kl_target:
+            self.kl_coeff_val *= 1.5
+        elif sampled_kl < 0.5 * self.kl_target:
+            self.kl_coeff_val *= 0.5
+        return self.kl_coeff_val
+
+    def postprocess_trajectory(self, sample_batch, other_agent_batches=None):
+        last_r = 0.0
+        batch = compute_advantages(
+            sample_batch, last_r, self.config["gamma"], self.config["lambda"])
+        return batch
+
+    def gradients(self, optimizer):
+        return optimizer.compute_gradients(
+            self._loss, colocate_gradients_with_ops=True)
+
+    def get_state(self):
+        return [TFPolicyGraph.get_state(self),
+                self.kl_target,
+                self.kl_coeff_val]
+
+    def set_state(self, state):
+        TFPolicyGraph.set_state(self, state[0])
+        self.kl_target = state[1]
```

Comment:
Note that if kl_coeff becomes part of the weights, then get/set state become unnecessary.